### PR TITLE
oelldb

### PR DESCRIPTION
**Pull request by anakrish**
Implement python extension for lldb to support debugging OE SGX enclaves.

Signed-off-by: Anand Krishnamoorthi <anakrish@microsoft.com>

### CODE REVIEW

File: debugger/lldb-extension/lldb_sgx_plugin.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        print ("Can't access memory at {0:x}.".format(int(addr)) + "\n" + str(os.error))
+        return None
+
+def read_int_from_memory(addr, size):
+    mv = read_from_memory(addr, size)
+    return int.from_bytes(mv, 'little')
+
+def target_path_to_host_path(target_path):
+    so_name = os.path.basename(target_path)
+    strpath = gdb.execute("show solib-search-path", False, True)
```
Comment (mingweishih) [inline debugger/lldb-extension/lldb_sgx_plugin.py:168]: Question: Are we supposed to use `gdb` here?

File: debugger/lldb-extension/lldb_sgx_plugin.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    if not enclave.is_valid():
+        return False
+
+    # No version specific checks.
+    # The contract will be extended in backwards compatible manner.
+    # Debugger may use version to take specific actions in future.
+
+    # Check if debugging is enabled.
+    if enclave.debug == 0:
+        print ("oegdb: Debugging not enabled for enclave %s" % enclave.path)
```
Comment (mingweishih) [inline debugger/lldb-extension/lldb_sgx_plugin.py:239]: Should it be `oelldb` instead? Here and the below.

File: debugger/lldb-extension/lldb_sgx_plugin.py (hunk 1 above)
Comment (anakrish) [reply debugger/lldb-extension/lldb_sgx_plugin.py:168]: Good catch. Will fix.

File: debugger/lldb-extension/lldb_sgx_plugin.py (hunk 2 above)
Comment (anakrish) [reply debugger/lldb-extension/lldb_sgx_plugin.py:239]: I need to do a string replace.

File: debugger/lldb-extension/lldb_sgx_plugin.py (hunk 1 above)
Comment (anakrish) [reply debugger/lldb-extension/lldb_sgx_plugin.py]: Fixed. It is not clear what scenario target_path_to_host_path addresses. I have removed it for oelldb.

**Comment by anakrish** [comment]
bors try

**Comment by dthaler** [comment]
This looks related to https://github.com/confidential-computing/governance/issues/51
I.e, the question is: is this really OE-specific?   Or can it be in a project that's reusable between OE, SGX SDK, Enarx, etc?

**Review by mingweishih — APPROVED** [review]

**Comment by anakrish** [comment]
> This looks related to [confidential-computing/governance#51](https://github.com/confidential-computing/governance/issues/51)
> I.e, the question is: is this really OE-specific? Or can it be in a project that's reusable between OE, SGX SDK, Enarx, etc?

Both `oegdb` and `oelldb` use a [sgx-enlightened ptrace library](https://github.com/openenclave/openenclave/tree/master/debugger/ptraceLib). The ask of [confidential-computing/governance#51](https://github.com/confidential-computing/governance/issues/51) is to upstream this ptrace library to glibc.

`oegdb` and `oelldb` are python extensions to `gdb` and `lldb` respectively. They are written using [OE SDK's debugger contract](https://github.com/openenclave/openenclave/tree/master/include/openenclave/internal/debugrt) and therefore should work for any SDK that implements the debugger contract.

SGX SDK for Windows implements the debugger contract. Therefore Visual Studio Debugger can be used to debug SGX SDK enclaves on Windows. I'm not sure if SGX SDK for Linux implements the debugger contract; if it does then both `oegdb` and `oelldb` can be used.

**Comment by anakrish** [comment]
> Or can it be in a project that's reusable between OE, SGX SDK, Enarx, etc?

By `project` are you referring to a separate repository? If so, yes the debuggers can be moved to a separate repository if we want to. We need to evaluate the benefits, CI costs etc.

**Review by radhikaj — APPROVED** [review]

**Comment by dthaler** [comment]
> > Or can it be in a project that's reusable between OE, SGX SDK, Enarx, etc?
> 
> By `project` are you referring to a separate repository? If so, yes the debuggers can be moved to a separate repository if we want to. We need to evaluate the benefits, CI costs etc.

yes I mean separate repo, so that it can be directly used by other projects besides OE, via a submodule.  This is an explicit ask from other CCC members.    CI costs can potentially ask for CCC support to cover if needed.

**Pull request merged**